module/os/linux/spl/spl-zone.c

```diff
@@ -0,0 +1,325 @@
+		return (error);
+
+	mutex_enter(&zone_datasets_lock);
+	zds = zone_datasets_lookup(nsnum);
+	if (zds == NULL) {
+		zds = kmem_alloc(sizeof (zone_datasets_t), KM_SLEEP);
+		INIT_LIST_HEAD(&zds->zds_list);
+		INIT_LIST_HEAD(&zds->zds_datasets);
+		zds->zds_nsnum = nsnum;
+		list_add_tail(&zds->zds_list, &zone_datasets);
+	}
+	zd = zone_dataset_lookup(zds, dataset, dsnamelen);
+	if (zd != NULL) {
+		error = EEXIST;
+		goto done;
+	}
+
+	zd = kmem_alloc(sizeof (zone_dataset_t) + dsnamelen + 1, KM_SLEEP);
+	zd->zd_dsnamelen = dsnamelen;
+	strncpy(zd->zd_dsname, dataset, dsnamelen);
+	zd->zd_dsname[dsnamelen] = '\0';
+	INIT_LIST_HEAD(&zd->zd_list);
+	list_add_tail(&zd->zd_list, &zds->zds_datasets);
+	error = 0;
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (error);
+#else
+	return (ENXIO);
+#endif
+}
+EXPORT_SYMBOL(zone_dataset_attach);
+
+int
+zone_dataset_detach(cred_t *cred, const char *dataset, unsigned int nsnum)
+{
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+	int error;
+	size_t dsnamelen;
+
+#if defined(CONFIG_USER_NS)
+	if ((error = zone_dataset_cred_check(cred)) != 0)
+		return (error);
+	if ((error = zone_dataset_name_check(dataset, &dsnamelen)) != 0)
+		return (error);
+
+	mutex_enter(&zone_datasets_lock);
+	zds = zone_datasets_lookup(nsnum);
+	if (zds != NULL)
+		zd = zone_dataset_lookup(zds, dataset, dsnamelen);
+	if (zds == NULL || zd == NULL) {
+		error = ENOENT;
+		goto done;
+	}
+
+	list_del(&zd->zd_list);
+	kmem_free(zd, sizeof (*zd) + zd->zd_dsnamelen + 1);
+
+	/* Prune the namespace entry if it has no more delegations. */
+	if (list_empty(&zds->zds_datasets)) {
+		list_del(&zds->zds_list);
+		kmem_free(zds, sizeof (*zds));
+	}
+	error = 0;
+
+done:
+	mutex_exit(&zone_datasets_lock);
+	return (error);
+#else
+	return (ENXIO);
+#endif
+}
+EXPORT_SYMBOL(zone_dataset_detach);
+
+int
+zone_dataset_visible(const char *dataset, int *write)
+{
+	zone_datasets_t *zds;
+	zone_dataset_t *zd;
+	int visible;
+	size_t dsnamelen, zd_len;
+
+	/* Default to read-only, in case visible is returned. */
+	if (write != NULL)
+		*write = 0;
+	if (zone_dataset_name_check(dataset, &dsnamelen) != 0)
+		return (0);
+	if (INGLOBALZONE(curproc)) {
+		if (write != NULL)
+			*write = 1;
+		return (1);
+	}
+
+	mutex_enter(&zone_datasets_lock);
+	visible = 0;
+	zds = zone_datasets_lookup(crgetzoneid(curproc->cred));
+	if (zds == NULL)
+		goto done;
+
+	list_for_each_entry(zd, &zds->zds_datasets, zd_list) {
+		zd_len = strlen(zd->zd_dsname);
+		/*
+		 * The dataset is visible only if it matches one of the
+		 * namespace's entries.  If it does, it's writable if it's
+		 * an exact match for or deeper than a namespace entry,
+		 * otherwise it's read-only.
+		 */
+		if (zd_len > dsnamelen) {
+			visible = bcmp(zd->zd_dsname, dataset,
+			    dsnamelen) == 0 &&
+			    zd->zd_dsname[dsnamelen] == '/';
+		} else {
+			visible = bcmp(zd->zd_dsname, dataset, zd_len) == 0;
+			if (visible && zd_len > dsnamelen)
```

Comment:
`zd_len > dsnamelen` is impossible here.

Reply:
When you do `zfs list`, it will call `zone_dataset_visible()` on each. so 'dataset' will be each dataset in the pool.

the delegated dataset is `containers/unpriv`
so we compare:
zd=containers/unpriv (17), ds=containers (10) visible=1
zd=containers/unpriv (17), ds=containers/unpriv (17) visible=1
zd=containers/unpriv (17), ds=containers/unpriv/child (23) visible=1
zd=containers/unpriv (17), ds=containers/unpriv/child/gchild (30) visible=1
zd=containers/unpriv (17), ds=containers/not-delegated-dataset (32) visible=0

Reply:
It's impossible because we wouldn't be in this block of code if the condition were true.
```C
if (zd_len > dsnamelen) {
    /* ... */
} else {
    /* implies !(zd_len > dsnamelen) */
}
```